source/common/tls/context_config_impl.cc

```
@@ -450,6 +451,27 @@ ServerContextConfigImpl::ServerContextConfigImpl(
     session_timeout_ =
         std::chrono::seconds(DurationUtil::durationToSeconds(config.session_timeout()));
   }
+
+  TlsCertificateSelectorFactoryContextImpl provider_factory_context(api_, options_,
+                                                                    singleton_manager_);
+  if (config.common_tls_context().has_custom_tls_certificate_selector()) {
+    // If a custom tls context provider is configured, derive the factory from the config.
+    const auto& provider_config = config.common_tls_context().custom_tls_certificate_selector();
+    Ssl::TlsCertificateSelectorFactory* provider_factory =
+        &Config::Utility::getAndCheckFactory<Ssl::TlsCertificateSelectorFactory>(provider_config);
+    tls_certificate_selector_factory_cb_ = provider_factory->createTlsCertificateSelectorCb(
+        provider_config.typed_config(), provider_factory_context,
+        factory_context.messageValidationVisitor());
+  } else {
+    auto factory = Envoy::Config::Utility::getFactoryByName<Ssl::TlsCertificateSelectorFactory>(
+        "envoy.ssl.certificate_selector_factory.default");
+    // mobile build doesn't have server_context.
+    if (factory) {
```

I think you should throw a validation exception if the factory isn't present.